dds/DCPS/DomainParticipantFactoryImpl.cpp

```diff
@@ -174,6 +174,14 @@ DomainParticipantFactoryImpl::delete_participant(
   DomainParticipantImpl* the_servant
   = dynamic_cast<DomainParticipantImpl*>(a_participant);
 
+  if (!the_servant) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: ")
+      ACE_TEXT("DomainParticipantFactoryImpl::delete_participant: ")
+      ACE_TEXT("failed to obtain the DomainParticipantImpl.\n")));
+
+    return DDS::RETCODE_PRECONDITION_NOT_MET;
```

Comment:
This might be better as simply a RETCODE_ERROR, since we should be able to dynamically cast the DomainParticipant_ptr, however it is also a precondition of being able to delete the participant that we have a conforming participant in the first place as well.